### PR TITLE
Https support

### PR DESCRIPTION
**Pull request by studioj**
as said in issue https://github.com/lensvol/intellij-blackconnect/issues/14

Adding a checkbox to enable the tool to run via https => default off

### CODE REVIEW

**Review by RaduBolcas — APPROVED** [review]
Looks good. Support for https is useful. It's a good option to avoid unencrypted code going back and forth which can introduce a security risk. 

Downloaded the patch and it worked for me

**Review by jasbeersingh — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -14,7 +14,7 @@ Instead, on each press of `Alt + Shift + B` plugin will send contents of the cur
 * Ability to reformat selected fragment only instead of whole file.
 * Load formatting settings from **[tool.black]** section of `pyproject.toml`
 * Configurable options:
-    * Hostname and port.
+    * Hostname,port and https.
```
Comment (crimsonxiii) [inline README.md:17]: space after comma

File: src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt [hunk 2]
```diff
@@ -14,12 +14,15 @@ sealed class BlackdResponse {
     data class UnknownStatus(val code: Int, val responseText: String) : BlackdResponse()
 }
 
-class BlackdClient(val hostname: String, val port: Int) {
+class BlackdClient(val hostname: String, val port: Int, val https: Boolean) {
```
Comment (crimsonxiii) [inline src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt:17]: https seems like an odd name to be honest
I would either expect protocol (a string that default has a value 'http') or a boolean like useSsl

File: src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt [hunk 3]
```diff
@@ -14,12 +14,15 @@ sealed class BlackdResponse {
...
 
     private val logger = Logger.getInstance(BlackdClient::class.java.name)
 
     fun checkConnection(): Result<String, String> {
-        val url = URL("https://$hostname:$port")
+        var url = URL("http://$hostname:$port")
+        if (https) {
```
Comment (crimsonxiii) [inline src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt:23]: val protocol = if (useSsl) "https" else "http
val url = URL("$protocol://$hostname:$port")

File: src/main/kotlin/me/lensvol/blackconnect/config/sections/ConnectionSection.kt [hunk 4]
```diff
@@ -18,6 +18,7 @@ import javax.swing.JButton
 import javax.swing.JComponent
 import javax.swing.JPanel
 import javax.swing.JSpinner
+import javax.swing.JCheckBox
```
Comment (crimsonxiii) [inline src/main/kotlin/me/lensvol/blackconnect/config/sections/ConnectionSection.kt:21]: sort

File: src/main/kotlin/me/lensvol/blackconnect/Constants.kt [hunk 5]
```diff
@@ -4,6 +4,7 @@ object Constants {
     const val DEFAULT_BLACKD_PORT = 45484
     const val DEFAULT_LINE_LENGTH = 88
     const val DEFAULT_HOST_BINDING = "localhost"
+    const val DEFAULT_HTTPS_ENABLED = false
```
Comment (crimsonxiii) [inline src/main/kotlin/me/lensvol/blackconnect/Constants.kt:7]: DEFAULT_USE_SSL if you rename the variable

**Review by crimsonxiii — CHANGES_REQUESTED** [review]

**Review by mirificus — APPROVED** [review]

File: src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt (hunk 3 above)
Comment (studioj) [reply src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt:23]: thx this Java Kotlin code is not my best friend :D didnt want to dive too deep. I'll adjust

**Comment by mirificus** [comment]
very nice PR,
should have been in there way earlier

File: src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt (hunk 2 above)
Comment (lensvol) [reply src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt:17]: I would agree, please consider naming this as something like `useSSL` or `connectOverHttps`. :)

File: src/main/kotlin/me/lensvol/blackconnect/config/sections/ConnectionSection.kt (hunk 4 above)
Comment (studioj) [reply src/main/kotlin/me/lensvol/blackconnect/config/sections/ConnectionSection.kt:21]: inteliJ suggested to do 

> import javax.swing.*

im a bit afraid to switch to * imports. In python I'm not really doing it eagerly

File: src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt (hunk 2 above)
Comment (studioj) [reply src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt]: I think keeping it as 'https' in the GUI seems most logical and user friendly.
@lensvol is the location ok for you? => how can i improve the space in between the boxes?
image

**Comment by studioj** [comment]
@lensvol is the look and feel of the settings menu still ok?

image

**Comment by studioj** [comment]
@lensvol the look and feel of the settings UI is still ok?

image

**Comment by lensvol** [comment]
@studioj I would recommend moving that checkbox under "Miscellaneous settings" and give it a longer description (e.g. "Connect to black[d] via HTTPS" :)

**Comment by lensvol** [comment]
@studioj Awesome work, by the way. :)

**Comment by studioj** [comment]
> @studioj I would recommend moving that checkbox under "Miscellaneous settings" and give it a longer description (e.g. "Connect to black[d] via HTTPS" :)

that would be harder since its needed to run the connection test

**Comment by lensvol** [comment]
@studioj Good point. Maybe it can be moved below the connection settings and above "Check connection" button?

I would still insist on longer description, tho :)

**Comment by studioj** [comment]
> @studioj Good point. Maybe it can be moved below the connection settings and above "Check connection" button?
> 
> I would still insist on longer description, tho :)
better? i dunno how to increase the space in between of the settings and the button => feels slightly better that way i would think
image

File: src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt [hunk 6]
```diff
@@ -51,8 +52,10 @@ class BlackdClient(val hostname: String, val port: Int) {
         skipStringNormalization: Boolean = false,
         targetPythonVersions: String = ""
     ): Result<BlackdResponse, Exception> {
-        val url = URL("http://$hostname:$port")
-
+        var url = URL("http://$hostname:$port")
+        if (useSsl) {
+            url = URL("https://$hostname:$port")
+        }
```
Comment (lensvol) [inline src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt:58]: At the moment, there are two approaches in the code that basically do the same thing. 

I suggest we merge them into one using property:

```
val blackdUrl: String
    get() =  URL("$protocol://$hostname:$port")
```

File: src/main/kotlin/me/lensvol/blackconnect/settings/BlackConnectProjectSettings.kt [hunk 7]
```diff
@@ -18,6 +18,9 @@ class BlackConnectProjectSettings : PersistentStateComponent<BlackConnectProject
             ServiceManager.getService(project, BlackConnectProjectSettings::class.java)
     }
 
+    @Attribute
+    var https: Boolean = Constants.DEFAULT_USE_SSL
```
Comment (lensvol) [inline src/main/kotlin/me/lensvol/blackconnect/settings/BlackConnectProjectSettings.kt:22]: Can `https` variable be renamed to something more descriptive like `useSSL` or `connectOverSSL`? As we would have to maintain that name for backward compatibility reasons. :)

File: src/test/kotlin/me/lensvol/blackconnect/BlackdClientTestCase.kt [hunk 8]
```diff
@@ -35,7 +35,7 @@ class BlackdClientTestCase : BlackConnectTestCase() {
             )
         }
 
-        val blackdClient = BlackdClient("localhost", 45484)
+        val blackdClient = BlackdClient("localhost", 45484, false)
```
Comment (lensvol) [inline src/test/kotlin/me/lensvol/blackconnect/BlackdClientTestCase.kt:38]: We can [pass the default value in the constructor](http://kotlin-quick-reference.com/102a-R-constructors-default-values.html) or do it explicitly at the instantiation:

```BlackdClient("localhost, 45484, useSSL = false)```

**Review by lensvol — CHANGES_REQUESTED** [review]
Thank you for the hard work! :) I've proposed some changes here and there, please consider if you would implement them and let me know.

File: src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt (hunk 6 above)
Comment (studioj) [reply src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt:58]: >     val blackdUrl: URL
        get() =  URL("$protocol://$hostname:$port")

you mean this as a property of BlackdClient

Comment (lensvol) [reply src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt:58]: Yes :)

Comment (studioj) [reply src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt:58]: private or public? => seems like public could deliver extra functionality ... later on (and allows me to write some extra tests :-)

**Review by lensvol — APPROVED** [review]

File: src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt (hunk 6 above)
Comment (lensvol) [reply src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt]: I think private should be enough for now.

**Comment by lensvol** [comment]
@studioj Thank you! :)

@crimsonxiii @jasbeersingh @RaduBolcas Kudos to you for doing the initial review :)

@mirificus Your opinion is duly noted.

File: src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt (hunk 6 above)
Comment (studioj) [reply src/main/kotlin/me/lensvol/blackconnect/BlackdClient.kt]: oops moved to public :-S

**Comment by studioj** [comment]
thx @lensvol for the quick resolution of the PR :-)
do you prefer me to close the fork?

**Comment by lensvol** [comment]
@studioj Up to you :)

**Comment by studioj** [comment]
@lensvol could you give an estimate when it will be on the pycharm plugin page?
Not pushing but to get an idea

**Comment by lensvol** [comment]
I cannot promise anything, but plan is to create a release this evening and then it usually takes 1-2 days to make it through the approval process.

**Comment by studioj** [comment]
Cool thx a lot.

Want me to clean up the connection message which has some not working html in it or was that done on purpose

**Comment by lensvol** [comment]
> Cool thx a lot.
> 
> Want me to clean up the connection message which has some not working html in it or was that done on purpose

What do you mean by 'not working HTML'?

**Comment by studioj** [comment]
> > Cool thx a lot.
> > Want me to clean up the connection message which has some not working html in it or was that done on purpose
> 
> What do you mean by 'not working HTML'?

image

**Comment by lensvol** [comment]
That is strange, I cannot reproduce it on my machine. :( 

If you can do it reliably and want to work on it, then feel free to submit another PR. It will be included in the next release.

**Comment by studioj** [comment]
het @lensvol  i cant connect anymore to the https server after yr changes

**Comment by studioj** [comment]
> hey @lensvol i cant connect anymore to the https server after yr changes
> 
> 
> ### also when using localhost it fails
> 

i solved it with assuming the TIMEOUTS are mili seconds instead of seconds

**Comment by lensvol** [comment]
@studioj You are right, that could be the reason and definitely would be fixed.

But I am curious. Could you possibly share details about your *blackd* setup? `Dockerfile`, specific versions of daemon and IDE etc. At the moment, I am having trouble replicating it and that can influence the outcome.

**Comment by studioj** [comment]
I can open a branch with an example docker file but I think I added the example I based myself on in the readme
 I don't know how the https implementation works exactly though. Most likely an nginx proxy
The docker is running remotely to provide the same black version for team members.
Using latest black[d] and latest black
Using the plugin with latest pycharm professional

**Pull request merged**